etl/src/energuide/transform.py

```diff
@@ -47,28 +51,44 @@ def extracted_rows(self) -> typing.Iterator[typing.Dict[str, typing.Any]]:
             for file in zip_input.namelist():
                 content = zip_input.read(file)
                 house = json.loads(content)
-                house['jsonFileName'] = content
+                house['jsonFileName'] = file
                 yield house
 
+    def num_rows(self) -> int:
+        with zipfile.ZipFile(self._zip_filename) as zip_input:
+            return len(zip_input.namelist())
+
 
 class AzureExtractReader:
 
     def __init__(self, coords: AzureCoordinates) -> None:
         self._coords = coords
+        self._azure: typing.Optional[blob.BlockBlobService] = None
 
-    def extracted_rows(self) -> typing.Iterator[typing.Dict[str, typing.Any]]:
-        azure_service = blob.BlockBlobService(account_name=self._coords.account,
-                                              account_key=self._coords.key,
-                                              custom_domain=self._coords.domain)
+    @property
+    def _azure_service(self) -> blob.BlockBlobService:
+        if self._azure is None:
+            self._azure = blob.BlockBlobService(account_name=self._coords.account,
+                                                account_key=self._coords.key,
+                                                custom_domain=self._coords.domain)
+        return self._azure
+
+    def _files(self) -> typing.List[str]:
         # itertools.groupby() needs its input sorted by the groupby key. We are assuming that that key is the filename
-        files = sorted([blob_.name for blob_ in azure_service.list_blobs(self._coords.container)
+        files = sorted([blob_.name for blob_ in self._azure_service.list_blobs(self._coords.container)
                         if 'timestamp' not in blob_.name])
-        for file in files:
-            content = azure_service.get_blob_to_bytes(self._coords.container, file).content
+        return files
+
+    def extracted_rows(self) -> typing.Iterator[typing.Dict[str, typing.Any]]:
+        for file in self._files():
+            content = self._azure_service.get_blob_to_bytes(self._coords.container, file).content
             house = json.loads(content)
-            house['jsonFileName'] = content
```

Comment:
This was the change that made the most difference to the messy output. On some errors, it was printing out the whole content instead of the filename.

Reply:
👍